kong/dns/utils.lua

```
@@ -0,0 +1,285 @@
+
+local LOCALHOST = {
+  ipv4 = "127.0.0.1",
+  ipv6 = "[::1]",
+}
+
+local DEFAULT_HOSTS = { localhost = LOCALHOST }
+
+
+local _M = {}
+
+
+-- checks the hostname type
+-- @return "ipv4", "ipv6", or "name"
+function _M.hostname_type(name)
+  local remainder, colons = name:gsub(":", "")
+  if colons > 1 then
+    return "ipv6"
+  end
+
+  if remainder:match("^[%d%.]+$") then
+    return "ipv4"
+  end
+
+  return "name"
+end
+
+
+-- parses a hostname with an optional port
+-- IPv6 addresses are always returned in square brackets
+-- @param name the string to check (this may contain a port number)
+-- @return `name/ip` + `port (or nil)` + `type ("ipv4", "ipv6" or "name")`
+function _M.parse_hostname(name)
+  local t = _M.hostname_type(name)
+  if t == "ipv4" or t == "name" then
+    local ip, port = name:match("^([^:]+)%:*(%d*)$")
+    return ip, tonumber(port), t
+  end
+
+  -- ipv6
+  if name:match("%[") then  -- brackets, so possibly a port
+    local ip, port = name:match("^%[([^%]]+)%]*%:*(%d*)$")
+    return "[" .. ip .. "]", tonumber(port), t
+  end
+
+  return "[" .. name .. "]", nil, t  -- no brackets also means no port
+end
+
+
+local function get_lines(path)
+  if type(path) == "table" then
+    return path
+  end
+  return readlines(path)
+end
+
+
+function _M.parse_hosts(path, enable_ipv6)
+  local lines, err = get_lines(path or DEFAULT_HOSTS_FILE)
+  if not lines then
+    log(NOTICE, "Invalid hosts file: ", err)
+    return DEFAULT_HOSTS
+  end
+
+  local hosts = {}
+
+  for _, line in ipairs(lines) do
+    -- Remove leading/trailing whitespaces and split by whitespace
+    local parts = {}
+    for part in line:gmatch("%S+") do
+      if part:sub(1, 1) == '#' then
+        break
+      end
+      table_insert(parts, part:lower())
+    end
+
+    -- Check if the line contains an IP address followed by hostnames
+    if #parts >= 2 then
+      local ip, _, family = _M.parse_hostname(parts[1])
+      if family ~= "name" then    -- ipv4/ipv6
```

`if family == "name"`, shall we raise a Lua error?

---

Linux ignores invalid lines in the hosts file.